File: libs/power-log-parser/src/lib/parsers/card-changed-parser.ts
import { GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { Action, ChangeEntity, Node } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class CardChangedParser implements ActionParser {
	readonly ParserName = 'CardChangedParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return stateType === StateType.PowerTaskList && node.Type === ChangeEntity;
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const changeEntity = node.Object as ChangeEntity;
		let eventName: string | null = null;
		if (this.GameState.CurrentEntities.get(changeEntity.Entity)!.GetTag(GameTag.ZONE) === (Zone.PLAY as number)) {
			eventName = 'CARD_CHANGED_ON_BOARD';
		} else if (
			this.GameState.CurrentEntities.get(changeEntity.Entity)!.GetTag(GameTag.ZONE) === (Zone.HAND as number)
		) {
			eventName = 'CARD_CHANGED_IN_HAND';
		} else if (
			this.GameState.CurrentEntities.get(changeEntity.Entity)!.GetTag(GameTag.ZONE) === (Zone.DECK as number)
		) {
			eventName = 'CARD_CHANGED_IN_DECK';
		}
		if (eventName == null) {
			return null;
		}
		const cardId = changeEntity.CardId;
		const entity = this.GameState.CurrentEntities.get(changeEntity.Entity)!;
		const controllerId = entity.GetEffectiveController();
		const creatorEntityId = changeEntity.GetTag(GameTag.CREATOR);
		const creatorEntityCardId = this.GameState.CurrentEntities.has(creatorEntityId)
			? this.GameState.CurrentEntities.get(creatorEntityId)!.CardId
			: null;
		let lastInfluencedByEntityId: number | null = null;
		let lastAffectedByCardId: string | null = null;
		if (node.Parent != null && node.Parent.Type === Action) {
			const parent = node.Parent.Object as Action;
			lastInfluencedByEntityId = parent?.Entity ?? null;
			lastAffectedByCardId =
				this.GameState.CurrentEntities.get(lastInfluencedByEntityId ?? 0)?.CardId ?? null;
		}
		return [
			GameEventProvider.Create(
				changeEntity.TimeStamp,
				eventName,
				GameEventHelper.CreateProvider(eventName, cardId, controllerId, entity.Id, this.StateFacade, {
					CreatorCardId: creatorEntityCardId,
					LastAffectedByEntityId: lastInfluencedByEntityId,
					LastAffectedByCardId: lastAffectedByCardId,
					Tags: entity.GetTagsCopy(),
				}),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/CardChangedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;
using System.Linq;

namespace HearthstoneReplays.Events.Parsers
{
    public class CardChangedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public CardChangedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(ChangeEntity);
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var changeEntity = node.Object as ChangeEntity;
            string eventName = null;
            if (GameState.CurrentEntities[changeEntity.Entity].GetTag(GameTag.ZONE) == (int)Zone.PLAY)
            {
                eventName = "CARD_CHANGED_ON_BOARD";
            }
            else if (GameState.CurrentEntities[changeEntity.Entity].GetTag(GameTag.ZONE) == (int)Zone.HAND)
            {
                eventName = "CARD_CHANGED_IN_HAND";
            }
            else if (GameState.CurrentEntities[changeEntity.Entity].GetTag(GameTag.ZONE) == (int)Zone.DECK)
            {
                eventName = "CARD_CHANGED_IN_DECK";
            }
            if (eventName == null)
            {
                return null;
            }
            var cardId = changeEntity.CardId;
            var entity = GameState.CurrentEntities[changeEntity.Entity];
            var controllerId = entity.GetEffectiveController();
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, null, null);
            var creatorEntityId = changeEntity.GetTag(GameTag.CREATOR);
            var creatorEntityCardId = GameState.CurrentEntities.ContainsKey(creatorEntityId)
                ? GameState.CurrentEntities[creatorEntityId].CardId
                : null;
            int? lastInfluencedByEntityId = null;
            string lastAffectedByCardId = null;
            if (node.Parent != null && node.Parent.Type == typeof(Action))
            {
                var parent = node.Parent.Object as Action;
                lastInfluencedByEntityId = parent?.Entity;
                lastAffectedByCardId = GameState.CurrentEntities.GetValueOrDefault(lastInfluencedByEntityId ?? 0)?.CardId;
            }
            return new List<GameEventProvider> { GameEventProvider.Create(
                changeEntity.TimeStamp,
                eventName,
                GameEvent.CreateProvider(
                    eventName,
                    cardId,
                    controllerId,
                    entity.Id,
                    StateFacade,
                    //gameState,
                    new {
                        CreatorCardId = creatorEntityCardId,
                        LastAffectedByEntityId = lastInfluencedByEntityId,
                        LastAffectedByCardId = lastAffectedByCardId,
                        Tags = entity.GetTagsCopy(),
                    }),
                true,
                node) };
        }
    }
}
